source/extensions/filters/http/ip_tagging/ip_tagging_filter.h

```diff
@@ -6,21 +6,141 @@
 
+#include "absl/hash/hash.h"
+#include "absl/strings/string_view.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace IpTagging {
 
+using IpTagFileProto = envoy::extensions::filters::http::ip_tagging::v3::IPTagging::IPTagFile;
+using IPTagsProto =
+    Protobuf::RepeatedPtrField<::envoy::extensions::filters::http::ip_tagging::v3::IPTagging_IPTag>;
+using TrieVector = std::vector<std::pair<std::string, std::vector<Network::Address::CidrRange>>>;
+using TagResolverFn =
+    std::function<std::vector<std::string>(const Network::Address::InstanceConstSharedPtr&)>;
+
+/**
+ * Class that manages setting all the stats for this filter.
+ */
+
+class IpTaggingFilterStats {
+
+public:
+  IpTaggingFilterStats(Stats::Scope& scope, const std::string& stat_prefix);
+
+  void incHit(absl::string_view tag) {
+    incCounter(stat_name_set_->getBuiltin(absl::StrCat(tag, ".hit"), unknown_tag_));
+  }
+
+  void ensureTagMetricExists(absl::string_view tag) {
+    stat_name_set_->rememberBuiltin(absl::StrCat(tag, ".hit"));
+  }
+
+  void incNoHit() { incCounter(no_hit_); }
+
+  void incTotal() { incCounter(total_); }
+
+  ~IpTaggingFilterStats();
+
+private:
+  void incCounter(Stats::StatName name);
+
+  Stats::Scope& scope_;
+  Stats::StatNameSetPtr stat_name_set_;
+  const Stats::StatName stats_prefix_;
+  const Stats::StatName no_hit_;
+  const Stats::StatName total_;
+  const Stats::StatName unknown_tag_;
+};
+
+/**
+ * Class that holds both stats and Trie together
+ */
+
+class LcTrieWithStats {
+
+public:
+  LcTrieWithStats(const TrieVector& trie, Stats::Scope& scope, const std::string& stats_prefix);
+
+  std::vector<std::string>
+  resolveTagsForIpAddress(const Network::Address::InstanceConstSharedPtr& Ipaddress);
+
+  ~LcTrieWithStats();
+
+private:
+  Network::LcTrie::LcTrie<std::string> trie_;
+  IpTaggingFilterStats filter_stats_;
+};
+
+/**
+ * Coordinates with the Filesystem::Watcher and when that reports a change, load up
+ * the change and updates it's internal settings.
+ */
+class TagSetWatcher {
+
+public:
+  static std::shared_ptr<TagSetWatcher>
+  create(Server::Configuration::FactoryContext& factory_context, std::string filename,
+         Stats::Scope& scope, const std::string& stat_prefix);
+
+  TagSetWatcher(Server::Configuration::FactoryContext& factory_context,
+                Event::Dispatcher& dispatcher, Api::Api& api, std::string filename,
+                Stats::Scope& scope, const std::string& stat_prefix);
+
+  TagSetWatcher(Server::Configuration::FactoryContext& factory_context, std::string filename,
+                Stats::Scope& scope, const std::string& stat_prefix)
+      : TagSetWatcher(factory_context, factory_context.dispatcher(), factory_context.api(),
+                      std::move(filename), scope, stat_prefix) {}
+
+  ~TagSetWatcher();
+
+  const std::string& filename() { return filename_; }
+
+  std::shared_ptr<LcTrieWithStats> get() const { return triestat_; }
```

Comment:
would LcTrieWithStats& be enough?

Reply:
Curious, what are we going with that change?

Reply:
I don't think you really need a shared ptr here, since the containing instance won't go out of scope right? Making a shared_ptr<> copy comes with a cost, plus it also makes assumptions about the lifetime of things. So in this case a reference might suffice, I think.

Reply:
Ok so this is called from:

```
    resolver_ = [watcherPtr](const Network::Address::InstanceConstSharedPtr& address) {
      return watcherPtr->get()->resolveTagsForIpAddress(address);
    };
```

so yeah I don't see the need for returning a shared ptr copy there, it should be enough to do:

```
    resolver_ = [watcherPtr](const Network::Address::InstanceConstSharedPtr& address) {
      return watcherPtr->get().resolveTagsForIpAddress(address);
    };
```

Reply:
Will make that change